### ISSUE TITLE
Android Wear support

### ISSUE THREAD

**Issue by flocke** [flocke/andOTP#68]
In the very distant future I would like to add an Android Wear app to display OTP tokens on a smartwatch.
It should only be able to display tokens, editing and adding should still only be able on the phone.

**Comment by flocke** [comment]
Yeah, I have never done anything for Wear as well. This is an ultra-longtime goal ...

**Comment by bovine3dom** [comment]
I'd like to see this feature and am happy to contribute some time towards it. I've never written any Java or done any android development before, so don't get too excited.

The first hurdle I see is communication between the mother device and the wearable: Google is quite keen to have you use Google Play Services for this (1), which I imagine is out of the question. The alternative is to just use Bluetooth (2), which should be fine as communication between the mothership and wearable would only be required when tokens are added/edited (so my dream of not having to have my phone on me whenever I log in to some service will be fulfilled).

There doesn't seem to be a wealth of information about making Wear applications without using Play Services, so this could be more interesting than I had first imagined.

1 - https://developer.android.com/training/wearables/data-layer/index.html
2 - https://developer.android.com/guide/topics/connectivity/bluetooth.html

**Comment by flocke** [comment]
While this is still on my TODO list, it is on indefinite hold for now, I want to clean up the codebase of andOTP first. Feel free to look into it but I assume it is gonna be a LOT of work.

The communication without Google Play Services would be extremely difficult but I at the same time don't see a point in trying to do this without them. As far as I know there is currently not really any way of running an Android Wear watch without the Play Services and the Android Wear app on the phone so we might as well use them. This of course would require us to provide a build flavor that excludes them (and with that won't support Android Wear) for F-Droid.

**Comment by f-ben** [comment]
Is there a new timeline for this? Your app is SO much better than google authenticator but the wear support ist really missing for me :(

**Comment by flocke** [comment]
Sorry, but the wear support is on ice for now. I don't even have enough time to maintain the main app in the moment, let alone write a new wear app.

**Comment by f-ben** [comment]
Thats sad to hear but thanks for the honest answer. I´ll try to get along without wear support then. Wish you all the best

**Comment by andrin55** [comment]
Did anything change on your time issue?

**Comment by flocke** [comment]
Not really, and even with more time there are still a lot of more important issues that need to be dealt with.

**Comment by Offerel** [comment]
Hopefully you find the time and resources to work on this topic. Sure, currently I have switched over to the F-droid variant of your app, but if it is required, I would change back to the Play Store (at least for this app).

But I completely understand the time problem. To bad that I can't help you with coding, that's completely out of my skills currently.

**Comment by DaneChainS** [comment]
Hello @Offerel (and maybe @andrin55 / @f-ben / @bovine3dom) 

Which watch do you have? You can beta test weaOTP  if it works for you on your wearOS watch (API >=26).
Either you transfer a backup file from andOTP to your watch and load it (better use encrypted, not plain) or you can use it via bluetooth with a modded andOTP version e.g. from the Galaxy Store (it has a different ID, so it does not interfere with the original app). 
I still need to write a guide how to use it, since it was only a personal use project until now, but if you're interested to test it, here a few hints:

- the backup file will be automatically transfered into .../Music/GalaxyWearable/ if you have a Galaxy Watch and use the wearable app (add music track to the watch). On other wearOS watches you have to create the folder on the watch and transfer the file there using Device File Explorer in Android Studio, a 3rd party file manager, or adb push.
- backup file name has to be accounts.json.aes.mp3 (for encrypted backups) or without .aes for plain files
- after you loaded some data, weaOTP will automatically show the codes when loaded. If you want to update your accounts or delete (wipe DB) them, press the back button to show the loading page. If you used the bluetooth link mode, the connection data is saved, so you only need to press "load" again.
- if you use the tile widget: there is a 20 second hard coded automatic update limiter in wearOS. So in the worst case if you swipe to the tile within 40 seconds twice, it might not automatically show the current codes. Just tap the middle to refresh in that case.
- tap on an icon in the tile to select the account to display there. After you did that, tap refresh to see changes immediatelely or the change will be visible after the next wearOS tile refresh cycle (max 20 seconds).

One more thing: SET A PIN CODE ON YOUR WATCH (so it will be locked when you take it off). 
I only implemented the things I personally use. So no HOTP counter-based accounts, only TOTP. 

When I have more time to clean up code and add functionality (and make it foolproof: "you were not supposed to do THAT"), maybe it can be added to andOTP in the future. But since it is only interesting for a small subset of users and might introduce new compability issues, maintainability effort etc., it might be better to keep it separate for now.

There is currently a beta update pending (not yet published on play store). I'll remove this note as soon as it is online in a couple days.

**Comment by bovine3dom** [comment]
Hi, I have a Huawei Watch 1. It's running Android Wear 2.35.0.333290778 - I don't know what API level that corresponds to :)

https://play.google.com/store/apps/details?id=com.andotp.weaotp says it's incompatible with my watch. I'd otherwise be delighted to spend a little time testing it

**Comment by DaneChainS** [comment]
@bovine3dom
I'm sorry, according to device database from google the Huawei Watch 1 is based on Android 7.1 (SDK 25). 
I think there was a specific reason why I needed to go up to minimum 26. I will look into it again what exactly the reason was... It was the tile-renderer library for the debug build only... there was a min 26 requirement.
But tiles already work on your watch, right?
I think I can work around that and the release version should maybe also work with 25... I will let you know when I have an update for you... Ok, at least the build already succeded and according to google it now supports 12 new devices including the Huawei Watch 1. It might take another week or so until it is verified.

**Comment by bovine3dom** [comment]
> But tiles already work on your watch, right?

Yes :)

There's no rush for me - I've been waiting patiently for three years so I can wait another week. Just ping me when you want me to test something!

**Comment by DaneChainS** [comment]
@bovine3dom  The new 25+ version should be online now. If it works at all for you, you can use / test it. Just let me know if you have problems or questions.
Following things might be interesting to test. But don't feel obligated to do that. 

Watch specific:
- Display size / readability / usability: Is any text not fitting / cut off in main app or tile?
- performance: is it fast & convenient for everyday usage? 
- how long does decrypting a backup file take? Which phone was used to create the andOTP backup file?
  - (Note regarding decryption performance: The PBKDF2 iterations in andOTP are benchmarked based on the device performance. So decryption/key generation on the watch will take a while (15 seconds for me) if your phone is much more powerful than your watch. Decryption of backup files uses quite a bit CPU power and battery. This is normal. But in every day usage only needed once or when you update the accounts. The every day watch internal keystore based DB encryption/decryption is much faster and easy on CPU and battery. So don't be worried when during testing you decrypt a lot of backup files and your battery life is less)
- every day battery life: after the initial testing just use it for while and see if the app and tile uses much extra battery.. (it should not... without loading new backups all the time ... see above).
- Eventual HW scrolling: if your watch has hardware scroll (like rotating bezel etc.) does it work in the app?
- any performance / visibility issues with excessive account count? Unlock the extended beta mode to have unlimited accounts displayed and have all experimental features activated. (Public version is limited until everything is stable)
- if you use the bluetooth sync functionality: how fast is it? Many crashes / transfer errors? (Most of the bluetooth issues are solved by completely closing (the modded) andOTP, reopen, unlock it and try again)

General:
- any issues regarding usability etc...? Wrong codes? Suggestions / improvements?
- performance of OnEnter tile code updates when main app is closed completely. To test it: app is closed and removed from background/recent. Wait at least 40 seconds. Better try it after couple of hours. Scroll to the tile. And see if it updates correctly and shows current codes (maybe have andOTP on the phone open to compare). I had the issue after long sleep time, OnEnter update sometimes did not work and I had to tap the refresh arrow in the tile.
Note: to preserve battery life only one update is triggered when swiping to the tile... there is no further updating if you stay on the tile (unless you entered it twice in the last 20sec). That's why the rest time is showed in the middle. You tap it to manually refresh.

You can also use the email address in the app for further and more specific feedback (to not bloat this issue thread more than necessary)

**Comment by JasonShallcross** [comment]
I've got the GalaxyWatch 4, it took me a while to get the file across via "add track", but got the eventually, took several tries.

Couldn't get the link option to work at all, just "Connection timeout", even though the app was open on the phone.

**Comment by DaneChainS** [comment]
If you want to spend time troubleshooting the link mode, please let me know [in this issue here](https://github.com/DaneChainS/weaOTP-doc/issues/3)
Currently only 3 people have been testing the app, and I'm the only one using link mode. (typical dev statement: Works fine on my device ;-) )
